### PR TITLE
Bidirectional token transfer

### LINKED ISSUE #696: Bi-directional token bridge for Litmus
### Context
This ticket could be an epic.
The current token bridge on Litmus is uni-directional (ETH => Litmus), we'd like to extend it to allow LIT transfer in the other direction. We could either use ChainSafe or Subbridge (from Phala) for it.

### Task
- [ ] investigate and decide which bridge solution to use (we are currently using chainSafe for ETH => Litmus)
- [ ] adjustment in code (pallet/runtime)
- [ ] ETH contract deployment and/or adjustment
- [ ] devops and configuration setup for the bridges
- [ ] Communicate with F/E team to test the bi-directional bridge

---
:heavy_check_mark: Please set appropriate **labels** and **assignees** if applicable.

### PR DESCRIPTION
**Pull request by ziming-zung**
resolves #696 

Run the integration test: `make test-ts-binary-rococo`

### CODE REVIEW

**Comment by Kailai-Wang** [comment]
Before I dive into this PR, have we tested it on testnet? Like rococo + rinkeby?

File: scripts/geth/run_geth.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
+# Copyright 2020 ChainSafe Systems
+# SPDX-License-Identifier: LGPL-3.0-only
```
Comment (Kailai-Wang) [inline scripts/geth/run_geth.sh]: Do we need this? Can we write something like "mostly copied from ...." even if it's copied?

File: Makefile [hunk 2]
```diff
@@ -121,6 +125,10 @@ test-ts-binary-litentry: launch-binary-litentry
 test-ts-binary-litmus: launch-binary-litmus
 	@./scripts/run-ts-test.sh
 
+test-ts-binary-rococo: launch-binary-rococo
```
Comment (Kailai-Wang) [inline Makefile]: Is it only used locally? As in CI I don't find any calls to `test-ts-binary-rococo`

File: scripts/launch-local-docker.sh [hunk 3]
```diff
@@ -32,6 +35,25 @@ parachain_service=$(docker-compose ps --services --filter 'status=running' | gre
...
+echo "copy binary:chainbridge to ${TMPDIR}"
+
+print_divider
+
+docker rm -f geth &>/dev/null
+docker run -d --rm --entrypoint 'sh' --name 'geth' -v ${ROOTDIR}/scripts/geth:/data/ -p 8546:8546 -p 8545:8545 ethereum/client-go:latest /data/run_geth.sh docker /data
+echo "runing geth...(container name: geth)"
+
+print_divider
+
```
Comment (Kailai-Wang) [inline scripts/launch-local-docker.sh]: I'll definitely move these to somewhere else (maybe a separate script), as this script is exclusively used to launch a parachain network for local testing => we normally don't want to run geth as well.

geth should only be run in CI or on demand.

File: ts-tests/bridge/contracts/MintableERC20.json [hunk 4]
```diff
@@ -0,0 +1,3909 @@
+{
```
Comment (Kailai-Wang) [inline ts-tests/bridge/contracts/MintableERC20.json:1]: 1. can we have a short README about how these contract JSONs are generated/copied from?
2. Are these the same as the production contracts?

File: scripts/build-docker.sh [hunk 5]
```diff
@@ -46,6 +46,10 @@ GITUSER=litentry
 GITREPO=litentry-parachain
 
 # Build the image
+echo "------------------------------------------------------------"
+echo "Building ${GITUSER}/chainbridge:latest docker image ..."
+docker build ${NOCACHE_FLAG} --pull -f ./docker/bridge.dockerfile -t ${GITUSER}/chainbridge:latest .
+
```
Comment (Kailai-Wang) [inline scripts/build-docker.sh]: I don't think it's necessary

File: scripts/launch-local-bridge-binary.sh [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+ROOTDIR=$(git rev-parse --show-toplevel)
+
+GOPATH=${HOME}/go go install github.com/litentry/ChainBridge/cmd/chainbridge@dev
```
Comment (Kailai-Wang) [inline scripts/launch-local-bridge-binary.sh:8]: It seems that `go install` is redundant? The binary is copied from docker image anyway

File: pallets/bridge/src/lib.rs [hunk 7]
```diff
@@ -448,6 +487,13 @@ pub mod pallet {
 			Self::chains(id).is_some()
 		}
 
+		/// Increments the deposit nonce for the specified chain ID
+		fn bump_nonce(id: BridgeChainId) -> DepositNonce {
+			let nonce = Self::chains(id).unwrap_or_default() + 1;
```
Comment (Kailai-Wang) [inline pallets/bridge/src/lib.rs]: Let's have `checked_add` here, although unlikely to overflow

File: pallets/bridge/src/lib.rs [hunk 8]
```diff
@@ -603,6 +649,50 @@ pub mod pallet {
...
+			ensure!(balance >= amount, Error::<T>::InsufficientBalance);
+
+			T::Currency::withdraw(&sender, actual_amount, WithdrawReasons::TRANSFER, AllowDeath)?;
+			T::Currency::burn(actual_amount);
+
+			// transfer fee to treasury
+			T::Currency::transfer(&sender, &T::TreasuryAccount::get(), fee, AllowDeath)?;
+
+			let nonce = Self::bump_nonce(dest_id);
+			BridgeEvents::<T>::append(BridgeEvent::FungibleTransfer(
```
Comment (Kailai-Wang) [inline pallets/bridge/src/lib.rs:688]: Where is `BridgeEvents` actually used? Is it for the relayers to query the storage and initiate contract transfer? 

If so can you please add a comment about it?

File: pallets/bridge/src/mock.rs [hunk 9]
```diff
@@ -98,14 +98,17 @@ impl pallet_balances::Config for Test {
 parameter_types! {
 	pub const TestChainId: u8 = 5;
 	pub const ProposalLifetime: u64 = 50;
+		pub const TreasuryAccount:u64 = 0x8;
```
Comment (Kailai-Wang) [inline pallets/bridge/src/mock.rs]: Small indention problem?

**Review by Kailai-Wang — COMMENTED** [review]
Thanks. The basic logic looks good to me, I have two extra questions:

- Is it feasible to enable the ts-test also in CI? So far I don't see how it's tested
- What if the transfer on ETH side goes wrong? We don't have any compensation or rollback mechanism, right?

File: scripts/geth/run_geth.sh (hunk 1 above)
Comment (ziming-zung) [reply scripts/geth/run_geth.sh]: In fact, this script is completely different from the original, we can delete these 2 lines

File: scripts/launch-local-docker.sh (hunk 3 above)
Comment (ziming-zung) [reply scripts/launch-local-docker.sh]: move these to a seperate file, please check.

File: ts-tests/bridge/contracts/MintableERC20.json (hunk 4 above)
Comment (ziming-zung) [reply ts-tests/bridge/contracts/MintableERC20.json:1]: 1. already added README.
2. answered in REAME, hope to explain clearly

File: scripts/build-docker.sh (hunk 5 above)
Comment (ziming-zung) [reply scripts/build-docker.sh]: already removed these code.

File: pallets/bridge/src/lib.rs (hunk 8 above)
Comment (ziming-zung) [reply pallets/bridge/src/lib.rs:688]: already added a comment, please check.

File: Makefile (hunk 2 above)
Comment (ziming-zung) [reply Makefile]: removed.

File: scripts/launch-local-bridge-binary.sh (hunk 6 above)
Comment (ziming-zung) [reply scripts/launch-local-bridge-binary.sh:8]: changed.

File: pallets/bridge/src/mock.rs (hunk 9 above)
Comment (ziming-zung) [reply pallets/bridge/src/mock.rs]: changed.

File: pallets/bridge/src/lib.rs (hunk 7 above)
Comment (ziming-zung) [reply pallets/bridge/src/lib.rs]: changed.

**Comment by ziming-zung** [comment]
> Thanks. The basic logic looks good to me, I have two extra questions:
> 
> * Is it feasible to enable the ts-test also in CI? So far I don't see how it's tested
> * What if the transfer on ETH side goes wrong? We don't have any compensation or rollback mechanism, right?

1. After modification, execute the command: ```./scripts/run-ts-test.sh bridge``` to run bridge test
2. After manual testing, I can answer you that there is no rollback or compensation mechanism when the cross-chain transfer fails. For more detailes, I will write a simple document to explain

File: scripts/geth/run_geth.sh (hunk 1 above)
Comment (Kailai-Wang) [reply scripts/geth/run_geth.sh]: Yes please :)

**Review by Kailai-Wang — APPROVED** [review]
Thanks, I approved this PR, as only some small things left:

- https://github.com/litentry/litentry-parachain/pull/740#discussion_r946702830
- I guess the contents of `scripts/launch-local-bridge-binary.sh` and `scripts/launch-local-bridge-docker.sh` are upside down
- do you want to activate token bridge on rococo right away? (Or let's do it in a separate PR)

**Pull request merged**